### PR TITLE
Плеханов Даниил. Задача 1. Вариант 21. Технология SEQ. Поиск кратчайших путей из одной вершины (алгоритм Дейкстры).

### PR DESCRIPTION
**Pull request by Dxppi**
# Алгоритм Дейкстры

Алгоритм Дейкстры предназначен для нахождения кратчайших путей от заданной начальной вершины до всех остальных вершин в графе.

## Представление графа

Для этого создается вектор списков смежности `adj_list`, где каждый элемент `adj_list[i]` представляет собой вектор пар `<dest, weight>`, указывающих на ребра, исходящие из вершины `i`. Данные о графе из `graph_data_` преобразуются в представление списка смежности. В этом формате данные о ребрах хранятся последовательно, а значение `-1` используется в качестве разделителя между списками ребер для разных вершин.

## Алгоритм Дейкстры

Алгоритм Дейкстры работает следующим образом:

1. **Инициализация**:
   - Создается вектор списков смежности `adj_list`, где каждый элемент представляет собой вектор пар `<dest, weight>`, указывающих на ребра, исходящие из вершины.
   - Для каждой вершины выделяется место для хранения предполагаемого количества ребер (в данном случае 8).

2. **Преобразование данных графа**:
   - Данные о графе из `graph_data_` преобразуются в представление списка смежности. Если встречается значение `kEndOfVertexList`, это означает, что список ребер для текущей вершины завершен, и переходим к следующей вершине.
   - Если вес ребра отрицательный, функция возвращает `false`, так как алгоритм Дейкстры не поддерживает отрицательные веса.

3. **Инициализация расстояний**:
   - Расстояние до начальной вершины `start_vertex_` устанавливается в 0.
   - Создается множество `vertex_set`, которое будет использоваться для хранения вершин, упорядоченных по текущему кратчайшему расстоянию.

4. **Основной цикл**:
   - Пока множество `vertex_set` не пусто, извлекается вершина `u` с наименьшим расстоянием.
   - Для каждого соседа `v` вершины `u` проверяется, можно ли улучшить текущее известное расстояние до `v`. Если новое расстояние меньше, чем текущее, оно обновляется, и вершина `v` добавляется в множество с новым расстоянием.

5. **Завершение**:
   - Алгоритм завершает свою работу, когда все вершины обработаны, и возвращает `true`.

### CODE REVIEW

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+#include <climits>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/plekhanov_d_dijkstra/include/ops_seq.hpp"
+
```
Comment (emilien-gus) [inline tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:12]: Please add validation test

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#include <gtest/gtest.h>
+
+#include <climits>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <utility>
+#include <vector>
+
```
Comment (OnePeaceEnjoyer) [inline tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:9]: Add tests with negative weights. It's shouldnt pass the validation test

Comment (Dxppi) [reply tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:9]: fixed, yeah, it doesn't pass the validation test

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp (hunk 1 above)
Comment (Dxppi) [reply tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:12]: fixed, added validation tests, works perfectly

**Review by emilien-gus — APPROVED** [review]

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+#include "core/task/include/task.hpp"
+#include "seq/plekhanov_d_dijkstra/include/ops_seq.hpp"
+namespace plekhanov_d_dijkstra_seq {
+
+void static RunValidationFailureTest();  // NOLINT
+
+template <typename ExpectedResultType>
+void RunTest(  // NOLINT
```
Comment (aobolensk) [inline tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:17]: What forced you to use NOLINT here? I guess it asks for anonymous namespace here

File: tasks/seq/plekhanov_d_dijkstra/perf_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    auto current_time_point = std::chrono::high_resolution_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(current_time_point - t0).count();
+    return static_cast<double>(duration) * 1e-9;
+  };
+
+  auto perf_results = std::make_shared<ppc::core::PerfResults>();
+
+  auto perf_analyzer = std::make_shared<ppc::core::Perf>(test_task_sequential);
+  perf_analyzer->PipelineRun(perf_attr, perf_results);
+  ppc::core::Perf::PrintPerfStatistic(perf_results);
```
Comment (aobolensk) [inline tasks/seq/plekhanov_d_dijkstra/perf_tests/main.cpp:68]: Please, add accuracy check in perf tests as well

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+#include <climits>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/plekhanov_d_dijkstra/include/ops_seq.hpp"
+namespace plekhanov_d_dijkstra_seq {
```
Comment (aobolensk) [inline tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:12]: Please, add random tests

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}  // namespace plekhanov_d_dijkstra_seq
+
+TEST(plekhanov_d_dijkstra_seq, test_dijkstra_Simple_Path_Graph) {
+  std::vector<std::vector<std::pair<size_t, int>>> adj_list = {
+      {{1, 1}}, {{0, 1}, {2, 2}}, {{1, 2}, {3, 3}}, {{2, 3}, {4, 4}}, {{3, 4}}};
+  std::vector<int> expected = {0, 1, 3, 6, 10};
+  plekhanov_d_dijkstra_seq::RunTest(adj_list, 0, expected);
+}
+
+TEST(plekhanov_d_dijkstra_seq, test_dijkstra_Complete_Graph) {
```
Comment (allnes) [inline tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:118]: please check differnet types of graph

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp (hunk 3 above)
Comment (Dxppi) [reply tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp]: fixed

File: tasks/seq/plekhanov_d_dijkstra/perf_tests/main.cpp (hunk 4 above)
Comment (Dxppi) [reply tasks/seq/plekhanov_d_dijkstra/perf_tests/main.cpp:103]: added accuracy check

File: tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp (hunk 5 above)
Comment (Dxppi) [reply tasks/seq/plekhanov_d_dijkstra/func_tests/main.cpp:16]: added random tests

**Review by allnes — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]
